### PR TITLE
Improves history request speed in backtest mode for python algorithms.

### LINKED ISSUE #2482: Improve History request speed in backtest mode for python algorithms.
#### Expected Behavior
In backtest mode, python algorithms should be able to read data files directly from disk to create a data frame.

#### Actual Behavior
Pandas data frames are created at C# side, relying on pythonnet. Global Interpreter Lock slow down the process.  

#### Potential Solution
Create a python layer algorithm in python with a History method to short-circuit the process.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by AlexCatarino**
#### Description
- Implements `QCAlgorithm.py`: serves as an interface between the python algorithms and the C# engine. The QCAlgorithm class implements a History method that overrides those from C# QCAlgorithm.
  - Updates python algorithms accordingly.
- Implements `PandasHistoryRequest` class. This class is used to gather the necessary information to manage the pandas data frame that is created by reading local files.
- Implements new methods in `QCAlgorithm.cs`: `CreateHistoryRequests(PyObject)`, `GetPandasHistoryRequests(PyObject)` and `PandasDataFrameHistory(PyObject)` that are used by the `History` method in `QCAlgorithm.py`.

#### Related Issue
Closes #2482 

#### Motivation and Context
Improve python algorithms speed.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
Added unit tests and regression tests.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by mchandschuh** [comment]
Reverted in fdc866fda04e3de5a2c4ec892a4569b3face5700

File: Common/Data/PandasHistoryRequest.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                    localDateTime = localDateTime.Add(Period);
+                }
+            }
+
+            DataType = IsCustomData ? config.Type : LeanData.GetDataType(resolution, TickType);
+
+            var namesByType = new Dictionary<Type, string[]>
+            {
+                { typeof(TradeBar), new[]{ "open", "hight", "low", "close", "volume" } },
+                { typeof(QuoteBar), new[]{ "bidopen", "bidhigh", "bidlow", "bidclose", "bidsize", "askopen", "askhigh", "asklow", "askclose", "asksize", } },
```
[the remark this answers is not in the document]
Comment (AlphaSquadAlgorithms) [reply Common/Data/PandasHistoryRequest.cs]: yes tick data types will be needed
quotebars.sliceData 
forex_market

File: Common/Data/PandasHistoryRequest.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+using QuantConnect.Util;
+
+namespace QuantConnect.Data
+{
+    /// <summary>
+    /// Represents a request for historical data using pandas
+    /// </summary>
+    public class PandasHistoryRequest : HistoryRequest
```
[the remark this answers is not in the document]
Comment (AlphaSquadAlgorithms) [reply Common/Data/PandasHistoryRequest.cs]: yes it seems to be required

File: Common/Data/PandasHistoryRequest.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    {
+        private static readonly IMapFileProvider _mapFileProvider = new LocalDiskMapFileProvider();
+        private static readonly IFactorFileProvider _factorFileProvider = new LocalDiskFactorFileProvider(_mapFileProvider);
```
[the remark this answers is not in the document]
Comment (AlphaSquadAlgorithms) [reply Common/Data/PandasHistoryRequest.cs]: suggest any changes?

File: Common/Data/PandasHistoryRequest.cs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+        /// <summary>
+        /// Date in the local file path
+        /// </summary>
+        public DateTime Date { get; }
```
[the remark this answers is not in the document]
Comment (AlphaSquadAlgorithms) [reply Common/Data/PandasHistoryRequest.cs]: these are data.sets with multiple request objects in one file

File: Algorithm.Python/Alphas/ForexCalendarAlpha.py [hunk 5]
```diff
@@ -14,18 +14,17 @@
...
 from QuantConnect import *
 from QuantConnect.Orders import *
-from QuantConnect.Algorithm import QCAlgorithm
 from datetime import timedelta, datetime
 from decimal import Decimal
 from Portfolio.EqualWeightingPortfolioConstructionModel import EqualWeightingPortfolioConstructionModel
 from Execution.ImmediateExecutionModel import ImmediateExecutionModel
 from Risk.NullRiskManagementModel import NullRiskManagementModel
 from QuantConnect.Data.Custom import DailyFx
+from QCAlgorithm import QCAlgorithm
```
[the remark this answers is not in the document]
Comment (AlphaSquadAlgorithms) [reply Algorithm.Python/Alphas/ForexCalendarAlpha.py]: suggestions as to which way should we procceed?

**Pull request merged**